File: 数据结构与算法/极客时间/二叉树/实现/BST.py
class Node:
    def __init__(self, data = None, left = None, right = None):
        self.data = data
        self.left = left
        self.right = right
    
    def show(self):
        return self.data

class BST:
    def __init__(self):
        self.head = Node()

    def insert(self, sourceNode):
        first = self.head;
        if first.data == None and first.left == None and first.right == None:
            self.head = sourceNode
        else:
            while True:
                if sourceNode.data < first.data:
                    if first.left == None:
                        first.left = sourceNode
                        break
                    else:
                        first = first.left
                else:
                    if first.right == None:
                        first.right = sourceNode
                        break
                    else:
                        first = first.right
    
    def preOrder(self, node):
        if node:
            print(node.show())
            self.preOrder(node.left)
            self.preOrder(node.right)
    
    def inOrder(self, node):
       if node:
            self.inOrder(node.left)
            print(node.show())
            self.inOrder(node.right)
    
    def postOrder(self, node):
       if node:
            self.postOrder(node.left)
            self.postOrder(node.right)
            print(node.show())
    
    def getMin(self) :
        first = self.head
        while first.left:
            first = first.left
        return first
    
    def getMax(self) :
        first = self.head
        while first.right:
            first = first.right
        return first


a = Node(1)
b = Node(2)
c = Node(3)
bst = BST()
bst.insert(a)
bst.insert(b)
bst.insert(c)
print(bst)
print(bst.preOrder(bst.head))

File: 数据结构与算法/极客时间/二叉树/实现/BST.js
function Node(data = null, left = null, right = null) {
  this.data = data;
  this.left = left;
  this.right = right;
  this.toString = function () {
    return this.data;
  };
}

function BST() {
  this.head = new Node();
}

BST.prototype.insert = function (sourceNode) {
  let first = this.head;
  if (first.data === null && first.left === null && first.right === null) {
    this.head = sourceNode;
  } else {
    while (true) {
      console.log(123, sourceNode.data, first.data);
      if (sourceNode.data > first.data) {
        /** 右节点 */
        if (first.right) {
          first = first.right;
        } else {
          first.right = sourceNode;
          break;
        }
      } else {
        /** 左节点 */
        if (first.left) {
          first = first.left;
        } else {
          first.left = sourceNode;
          break;
        }
      }
    }
  }
};

/** 前序: 根 -> 左 -> 右 */
BST.prototype.preOrder = function (node) {
  if (node) {
    console.log(node.toString());
    this.preOrder(node.left);
    this.preOrder(node.right);
  }
};

/** 中序: 左 -> 根 -> 右 */
BST.prototype.inOrder = function () {
  if (node) {
    this.inOrder(node.left);
    console.log(node.toString());
    this.inOrder(node.right);
  }
};

/** 后序: 左 -> 右 -> 根 */
BST.prototype.postOrder = function () {
  if (node) {
    this.postOrder(node.left);
    this.postOrder(node.right);
    console.log(node.toString());
  }
};

/** 最大 */
BST.prototype.getMax = function () {
  let first = this.head;
  while (true) {
    if (first.right) {
      first = first.right;
    } else {
      return first.data;
    }
  }
};

/** 最小 */
BST.prototype.getMin = function () {
  let first = this.head;
  while (true) {
    if (first.left) {
      first = first.left;
    } else {
      return first.data;
    }
  }
};

var a = new Node(1);
var b = new Node(2);
var c = new Node(3);
var list = new BST();
list.insert(a);
list.insert(b);
list.insert(c);
console.log(list);
console.log(list.preOrder(list.head));
